### PR TITLE
Add a view that emulates the SYS_CALENDAR.CALENDAR table from Teradata

### PR DESCRIPTION
**Pull request by Hackerpilot**
This will help people migrating code from Teradata to BigQuery. Placing this in views/migration/teradata to mimic the structure of the UDF directories. So far there seems to be no test infrastructure for views, so let me know what, if anything, needs to be done to test this. So far this has just been manually tested.

### CODE REVIEW

File: views/migration/teradata/sys_calendar.sql [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    quarter_of_calendar,
+    /* 1-n the year */
+    year_of_calendar
+  ) AS SELECT
+    r AS calendar_date,
+    EXTRACT(DAYOFWEEK FROM r) AS day_of_week,
+    EXTRACT(DAY FROM r) as day_of_month,
+    EXTRACT(DAYOFYEAR FROM r) AS day_of_year,
+    DATE_DIFF(r, DATE(1, 1, 1), DAY) AS day_of_calendar,
+    MOD(EXTRACT(DAYOFWEEK FROM DATE(
```
[the remark this answers is not in the document]
Comment (shevek-google) [reply views/migration/teradata/sys_calendar.sql:57]: It's (julian_day(d) - julian_day(some useful sunday)) / 7 - the question is can we compute the useful sunday more elegantly than this.

**Review by shevek-google — APPROVED** [review]

**Comment by Hackerpilot** [comment]
Pushed a new version that fixes a few cases. Tested with this:

```sql
select * from 
(select 4, weekday_of_month,
  4, week_of_month, calendar_date
  from dummy.CALENDAR
  where calendar_date = '2022-10-24')
union all
(select 3, weekday_of_month,
  3, week_of_month, calendar_date
  from dummy.CALENDAR
  where calendar_date = '2022-10-18')
union all
(select 5, weekday_of_month,
  4, week_of_month, calendar_date
  from dummy.CALENDAR
  where calendar_date = '2022-10-29')
union all
(select 1,  weekday_of_month,
  0, week_of_month, calendar_date
  from dummy.CALENDAR
  where calendar_date = '2022-10-1')
union all
(select 1, weekday_of_month,
  1, week_of_month, calendar_date
  from dummy.CALENDAR
  where calendar_date = '2022-5-1')
  union all
(select 1, weekday_of_month,
  0, week_of_month, calendar_date
  from dummy.CALENDAR
  where calendar_date = '2022-9-1')
  union all
(select 1, weekday_of_month,
  0, week_of_month, calendar_date
  from dummy.CALENDAR
  where calendar_date = '2022-9-2')
  union all
(select 1, weekday_of_month,
  0, week_of_month, calendar_date
  from dummy.CALENDAR
  where calendar_date = '2022-9-3')
    union all
(select 2, weekday_of_month,
  1, week_of_month, calendar_date
  from dummy.CALENDAR
  where calendar_date = '2022-9-9');
```

File: views/migration/teradata/README.md [hunk 2]
```diff
@@ -0,0 +1,22 @@
+# Teradata Views
+
+This directory contains views that mimic the behavior of tables / views in Teradata.
+
+## SYS_CALENDAR.CALENDAR
```
Comment (danieldeleo) [inline views/migration/teradata/README.md:5]: Add link to the view file

Comment (danieldeleo) [reply views/migration/teradata/README.md:5]: Also add a description about the view and mention that the columns of the view are described below

File: views/migration/teradata/sys_calendar.sql [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    /* 1-7 day of week */
+    day_of_week,
+    /* 1-n day of month */
+    day_of_month,
+    /* 1-365 day of year */
+    day_of_year,
+    /* 1-n day since the beginning of the calendar */
+    day_of_calendar,
+    /* 1 if the date is a Monday and the first Monday of the month, etc. */
+    weekday_of_month,
```
Comment (franklinWhaite) [inline views/migration/teradata/sys_calendar.sql:30]: Could you please explain a bit more what `weekday_of_month` is maybe add more detail to the description?

File: views/migration/teradata/sys_calendar.sql [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    DATE_DIFF(r, DATE(1, 1, 1), DAY) AS day_of_calendar,
+    MOD(EXTRACT(DAYOFWEEK FROM DATE(
+      EXTRACT(YEAR FROM r),
+      EXTRACT(MONTH FROM R),
+      1)), 7) + CAST(FLOOR((EXTRACT(DAY FROM r) - 1) / 7) AS INT64) + 1
+      AS weekday_of_month,
+    EXTRACT(WEEK from r) - EXTRACT(WEEK FROM DATE(
+      EXTRACT(YEAR FROM r),
+      EXTRACT(MONTH FROM R),
+      1)) AS week_of_month,
```
[the remark this answers is not in the document]
Comment (Hackerpilot) [reply views/migration/teradata/sys_calendar.sql]: This now uses date_trunc.

File: views/migration/teradata/README.md (hunk 2 above)
Comment (Hackerpilot) [reply views/migration/teradata/README.md:5]: Done.

File: views/migration/teradata/sys_calendar.sql (hunk 3 above)
Comment (Hackerpilot) [reply views/migration/teradata/sys_calendar.sql:32]: I've added a bit more to the comment here and in the README.

File: views/migration/teradata/sys_calendar.sql [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    quarter_of_calendar,
+    /* 1-n the year */
+    year_of_calendar
+  ) AS SELECT
+    r AS calendar_date,
+    EXTRACT(DAYOFWEEK FROM r) AS day_of_week,
+    EXTRACT(DAY FROM r) as day_of_month,
+    EXTRACT(DAYOFYEAR FROM r) AS day_of_year,
+    DATE_DIFF(r, DATE(1, 1, 1), DAY) AS day_of_calendar,
+    CAST(FLOOR(EXTRACT(DAY FROM r) / 7) AS INT64) + 1 AS weekday_of_month,
```
Comment (franklinWhaite) [inline views/migration/teradata/sys_calendar.sql:59]: This value is returning `2` for Friday Oct 7th (2022-10-07). It should be a `1` I believe

Comment (Hackerpilot) [reply views/migration/teradata/sys_calendar.sql]: Fixed.

File: views/migration/teradata/sys_calendar.sql [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    day_of_week,
+    /* 1-n day of month */
+    day_of_month,
+    /* 1-365 day of year */
+    day_of_year,
+    /* 1-n day since the beginning of the calendar */
+    day_of_calendar,
+    /* 1 if the date is a Monday and the first Monday of the month, 2 if it's a
+	 * Monday and the second Monday of the month, 4 for the fourth Tuesday, etc.
+	 */
```
Comment (franklinWhaite) [inline views/migration/teradata/sys_calendar.sql:31]: nit. this two lines seem to have and extra indentation.

File: views/migration/teradata/sys_calendar.sql [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    /* 1-n the year */
+    year_of_calendar
+  ) AS SELECT
+    r AS calendar_date,
+    EXTRACT(DAYOFWEEK FROM r) AS day_of_week,
+    EXTRACT(DAY FROM r) as day_of_month,
+    EXTRACT(DAYOFYEAR FROM r) AS day_of_year,
+    DATE_DIFF(r, DATE(1, 1, 1), DAY) AS day_of_calendar,
+    CAST(FLOOR((EXTRACT(DAY FROM r) - 1) / 7) AS INT64) + 1 AS weekday_of_month,
+    EXTRACT(WEEK FROM r) - EXTRACT(WEEK FROM DATE_TRUNC(r, MONTH)) + IF(EXTRACT(DAYOFWEEK FROM DATE_TRUNC(r, MONTH)) = 1, 1, 0) AS week_of_month,
```
Comment (franklinWhaite) [inline views/migration/teradata/sys_calendar.sql:60]: extreme nit. per style guide a maximum line length of 100 characters is enforced, this line is 145

File: views/migration/teradata/sys_calendar.sql (hunk 6 above)
Comment (Hackerpilot) [reply views/migration/teradata/sys_calendar.sql]: My editor used tabs by mistake. Fixed

File: views/migration/teradata/sys_calendar.sql (hunk 7 above)
Comment (Hackerpilot) [reply views/migration/teradata/sys_calendar.sql]: Fixed.

**Review by franklinWhaite — APPROVED** [review]
LGTM

**Pull request merged**